sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonFilters.scala

```diff
@@ -0,0 +1,131 @@
+    if (SQLConf.get.jsonFilterPushDown) {
+      val groupedByRefSet = filters
+        // Group filters that have the same set of references. For example:
+        //   IsNotNull("i") -> Set("i"), AlwaysTrue -> Set(),
+        //   And(EqualTo("i", 0), StringStartsWith("s", "abc")) -> Set("i", "s")
+        // By grouping filters we could avoid tracking their state of references in the
+        // current row separately.
+        .groupBy(_.references.toSet)
+        // Combine all filters from the same group by `And` because all filters should
+        // return `true` to do not skip a row. The result is compiled to a predicate.
+        .map { case (refSet, refsFilters) =>
+          (refSet, JsonPredicate(toPredicate(refsFilters), refSet.size, 0))
+        }
+      // Apply predicates w/o references like AlwaysTrue and AlwaysFalse to all fields.
+      // We cannot set such predicates to a particular position because skipRow() can
+      // be invoked for any index due to unpredictable order of JSON fields in JSON records.
+      val withLiterals = groupedByRefSet.map { case (refSet, pred) =>
+        if (refSet.isEmpty) {
+          (schema.fields.map(_.name).toSet, pred.copy(totalRefs = 1))
+        } else {
+          (refSet, pred)
+        }
+      }
+      // Build a map where key is only one field and value is seq of predicates refer to the field
+      // "i" -> Seq(AlwaysTrue, IsNotNull("i"), And(EqualTo("i", 0), StringStartsWith("s", "abc")))
+      // "s" -> Seq(AlwaysTrue, And(EqualTo("i", 0), StringStartsWith("s", "abc")))
+      val groupedByFields = withLiterals.toSeq
+        .flatMap { case (refSet, pred) => refSet.map((_, pred)) }
+        .groupBy(_._1)
+      // Build the final array by converting keys of `groupedByFields` to their
+      // indexes in the provided schema.
+      groupedByFields.foreach { case (fieldName, fieldPredicates) =>
+        val fieldIndex = schema.fieldIndex(fieldName)
+        groupedPredicates(fieldIndex) = fieldPredicates.map(_._2).toArray
+      }
+    }
+    groupedPredicates
+  }
+
+  /**
+   * Applies predicates (compiled filters) associated with the row field value
+   * at the position `index` only if other predicates dependencies are already
+   * set in the given row.
+   *
+   * @param row The row with fully or partially set values.
+   * @param index The index of already set value.
+   * @return true if at least one of applicable predicates (all dependent row values are set)
+   *         return `false`. It returns `false` if all predicates return `true`.
+   */
+  def skipRow(row: InternalRow, index: Int): Boolean = {
+    var skip = false
+    predicates(index).foreach { pred =>
+      pred.refCount -= 1
+      if (!skip && pred.refCount == 0) {
+        skip = !pred.predicate.eval(row)
+      }
+    }
+    skip
+  }
+
+  /**
+   * Reset states of all predicates by re-initializing reference counters.
+   */
+  override def reset(): Unit = predicates.foreach(_.foreach(_.reset))
```

Comment:
Actually, we cannot avoid having the state in the json filters? Is that technically difficult?

Reply:
I do believe it is impossible to avoid a state because JSON fields can appear in any order. So, to know when a filter can be applied, we need to keep a state w/ reference counter.

Reply:
I have an idea. I will maybe try it in a followup and see if it looks good all of you.